### PR TITLE
[Kogito 8699] Deploy the Operator on the New Cluster

### PR DESCRIPTION
**Pull request by desmax74**
See: https://issues.redhat.com/browse/KOGITO-8699

**Description of the change:**

**Motivation for the change:**

**Checklist**

- [ ] Add or Modify a unit test for your change
- [ ] Have you verified that tall the tests are passing?

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

### CODE REVIEW

**Comment by davidesalerno** [comment]
Isn't it better to followthe current conventions and put everythin into the .ci/tekton folder?

**Comment by desmax74** [comment]
> Isn't it better to follow the current conventions and put everythin into the .ci/tekton folder?

It's mainly documentation, not scripts to be executed on github

**Comment by spolti** [comment]
I agree with @davidesalerno, better to put it into a specific directory.

File: ci/PIPELINE.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
+## Install and configure a Tekton pipeline on Openshift
+
+1) Install Red Hat OpenShift Pipelines latest (1.9.2 at the time of this doc)  from the Operator Hub
```
Comment (spolti) [inline ci/PIPELINE.md:3]: Is there a official doc to link here?

File: ci/PIPELINE.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+2) If you want interact via cli with the pipeline you can install locally Tekton cli 
+   to interact with the pipelines, you can download from the details of the operator installed (i.e. crc link) 
+   https://tkn-cli-serve-openshift-pipelines.apps-crc.testing/tkn/tkn-linux-amd64.tar.gz
+   The version proposed by the Operator is correctly aligned version with the tekton version.
+
+
+3) If isn't yet created create the project kogito-serverless-operator-system
```
Comment (spolti) [inline ci/PIPELINE.md:11]: if what is not created?

Comment (spolti) [inline ci/PIPELINE.md:11]:
```suggestion
3) If isn't yet created, create the project kogito-serverless-operator-system
```

File: ci/PIPELINE.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+oc new-project kogito-serverless-operator-system
+```
+
+4) Install the tekton task called kubernetes-actions
+```sh
+kubectl apply -f https://api.hub.tekton.dev/v1/resource/tekton/task/kubernetes-actions/0.2/raw
+```
+5) Apply the cluster_role and cluster_role_binding
+```sh
+kubectl create -f ci/role/cluster_role.yaml
```
Comment (spolti) [inline ci/PIPELINE.md:22]:
```suggestion
kubectl create -f .ci/role/cluster_role.yaml
```

File: ci/PIPELINE.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+kubectl create -f ci/role/cluster_role_binding.yaml
```
Comment (spolti) [inline ci/PIPELINE.md:23]:
```suggestion
kubectl create -f .ci/role/cluster_role_binding.yaml
```

**Comment by spolti** [comment]
+1 for `.ci/tekton`

File: ci/PIPELINE.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
+## Install and configure a Tekton pipeline on Openshift
+
+1) Install Red Hat OpenShift Pipelines latest (1.9.2 at the time of this doc)  from the Operator Hub
+
+2) If you want interact via cli with the pipeline you can install locally Tekton cli 
+   to interact with the pipelines, you can download from the details of the operator installed (i.e. crc link) 
+   https://tkn-cli-serve-openshift-pipelines.apps-crc.testing/tkn/tkn-linux-amd64.tar.gz
+   The version proposed by the Operator is correctly aligned version with the tekton version.
```
Comment (radtriste) [inline ci/PIPELINE.md:8]:
```suggestion
2) If you want interact via cli with the pipeline you can install locally Tekton cli  
   To interact with the pipelines, you can download from the details of the operator installed (i.e. crc link):  
   https://tkn-cli-serve-openshift-pipelines.apps-crc.testing/tkn/tkn-linux-amd64.tar.gz  
   The version proposed by the Operator is correctly aligned version with the tekton version.
```

File: ci/PIPELINE.md (hunk 2 above)
Comment (radtriste) [inline ci/PIPELINE.md:11]:
```suggestion
3) If isn't yet created create the project `kogito-serverless-operator-system`
```

File: ci/PIPELINE.md (hunk 3 above)
Comment (radtriste) [reply ci/PIPELINE.md:22]: nope folder is called `ci` for now

File: ci/PIPELINE.md (hunk 4 above)
Comment (radtriste) [reply ci/PIPELINE.md:23]: same here, `ci` folder

File: ci/PIPELINE.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+```
+
+7) Run the pipeline creating a pipeline run 
+```sh
+kubectl apply -f ci/pipeline/kogito_serverless_operator_pipeline_run.yaml
+```
+
+or with tekton cli
+
+or with tekton cli
```
Comment (radtriste) [inline ci/PIPELINE.md:38]:
```suggestion

or with tekton cli:
```

File: ci/operator.yaml [hunk 7]
```diff
@@ -0,0 +1,3028 @@
+apiVersion: v1
```
Comment (radtriste) [inline ci/operator.yaml:1]: Please do not copy the operator.yaml here ...
please link to https://github.com/kiegroup/kogito-serverless-operator/blob/main/operator.yaml directly and remove this file.
We don't want to have to manage 2 times the same file in different places

File: ci/PIPELINE.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+ 
+Open the Pipeline Menu under the namespace/project kogito-serverless-operator-system
+or with tekton cli with the id printed starting the pipelinerun from the cli
+
+```sh
+tkn pipelinerun logs kogito-serverless-operator-pipeline-run-<id> -f -n <your-namespace>
+```
+
+### How to see the content of the workspace
+1) Create the task show_workspace_content
```
Comment (radtriste) [inline ci/PIPELINE.md:59]:
```suggestion
### How to see the content of the workspace

1) Create the task show_workspace_content
```

File: ci/PIPELINE.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+```sh
+kubectl apply -f ci/task/show_workspace_content.yaml
+```
+
+2) Add the task show-workspace in the pipeline after the fetch-repository or build-image
+
+### How to deploy again 
+Go to the pipeline runs and ask on a already executed pipelinerun a rerun
```
Comment (radtriste) [inline ci/PIPELINE.md:67]:
```suggestion
### How to deploy again 

Go to the pipeline runs and ask a rerun on an already executed pipeline
```

Comment (radtriste) [reply ci/PIPELINE.md:67]: btw, how to do that ?

File: ci/PIPELINE.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+NAME                                            TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)             AGE
+service/el-kogito-serverless-operator-webhook   ClusterIP   10.217.5.248   <none>        8080/TCP,9000/TCP   17m
+```
+
+7) Create Event Listener Ingress
+```sh
+   oc expose svc -n kogito-serverless-operator-pipeline el-kogito-swf-operator-webhook
+   export EL_WEBHOOK_HOST=$(oc get route el-gitea-webhook -n triggers-demo -ojsonpath='{.spec.host}')
+```
+8) Add a webhook in your github/gitlab repo with the url of the listener on openshift
```
Comment (radtriste) [inline ci/PIPELINE.md:117]: no webhook secret ?
I would suggest it for security reasons

**Comment by radtriste** [comment]
As pointed out by Davide and Filippe, I would also move that documentation to another folder.
What about `.ci/docs/` ?

File: ci/operator.yaml (hunk 7 above)
Comment (desmax74) [reply ci/operator.yaml:1]: @radtriste isnt' the same this is the line needed to deploy the image in the registry https://github.com/kiegroup/kogito-serverless-operator/pull/93/files#diff-0f6092c791b7d9467439876ba331d3043b4f3f6d66843fa766c4a81ebce4df9aR2997

File: ci/PIPELINE.md (hunk 10 above)
Comment (desmax74) [reply ci/PIPELINE.md:117]: @radtriste  Do we want to enable the triggers in a internal cluster ?

File: ci/PIPELINE.md (hunk 9 above)
Comment (desmax74) [reply ci/PIPELINE.md:67]: Is available a link on the pipelinerun executed

File: ci/PIPELINE.md (hunk 1 above)
Comment (desmax74) [reply ci/PIPELINE.md:3]: IS available this but is tied to a specific version of openshift https://docs.openshift.com/container-platform/4.12/cicd/pipelines/installing-pipelines.html

File: ci/PIPELINE.md (hunk 2 above)
Comment (desmax74) [reply ci/PIPELINE.md:11]: the namespace/project

File: ci/PIPELINE.md (hunk 10 above)
Comment (desmax74) [reply ci/PIPELINE.md:117]: Ok, I'll add this section of the doc https://docs.openshift.com/container-platform/4.12/cicd/pipelines/securing-webhooks-with-event-listeners.html

File: ci/operator.yaml (hunk 7 above)
Comment (radtriste) [reply ci/operator.yaml:1]: I see you use it here: https://github.com/kiegroup/kogito-serverless-operator/pull/93/files#diff-14c4d45418ce873b2d2ec357d9d66b7e813d4ff11de6bec43202421143b37027R55
Then please bring on a bash command there to take the file https://github.com/kiegroup/kogito-serverless-operator/blob/main/operator.yaml, update the image and apply the modified content

Could be something like:
```
curl https://raw.githubusercontent.com/kiegroup/kogito-serverless-operator/main/operator.yaml | sed -E 's|image: quay.io/kiegroup/.*|image: image-registry.openshift-image-registry.svc:5000/kogito-serverless-operator-system/kogito-serverless-operator:latest|g' | kubectl apply -f -
```

File: ci/PIPELINE.md (hunk 9 above)
Comment (radtriste) [reply ci/PIPELINE.md:67]: on the UI you mean ?

File: ci/PIPELINE.md (hunk 2 above)
Comment (desmax74) [reply ci/PIPELINE.md:11]: ok

File: ci/PIPELINE.md (hunk 5 above)
Comment (desmax74) [reply ci/PIPELINE.md:8]: ok

File: ci/PIPELINE.md (hunk 6 above)
Comment (desmax74) [reply ci/PIPELINE.md:38]: ok

File: ci/PIPELINE.md (hunk 9 above)
Comment (desmax74) [reply ci/PIPELINE.md:67]: Yep

**Comment by desmax74** [comment]
@radtriste if you clone the repo every folder named with a dot in the name is hidden, not a great solution to move the documentation in a hidden folder

File: ci/PIPELINE.md (hunk 2 above)
Comment (desmax74) [reply ci/PIPELINE.md]: ok

File: docs/PIPELINE.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+### How to see the content of the workspace
+
+1) Create the task show_workspace_content
+
+```sh
+kubectl apply -f tekton/task/show_workspace_content.yaml
+```
+
+2) Add the task show-workspace in the pipeline after the fetch-repository or build-image
```
Comment (radtriste) [inline docs/PIPELINE.md:64]:
```suggestion
2) Add the task `show-workspace` in the pipeline after the `fetch-repository` or `build-image`
```

File: docs/PIPELINE.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+### How to deploy again 
+
+Go to the pipeline runs and ask on a already executed pipelinerun a rerun
```
Comment (radtriste) [inline docs/PIPELINE.md:68]:
```suggestion
Go to the pipeline runs and ask for a rerun of a previous pipeline run
```

File: docs/PIPELINE.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+
+## Trigger the pipeline on GithubEvents
+
+1) Create the trigger binding
+```sh
+oc create -f tekton/trigger/trigger_binding.yaml
+```
+2) Create the trigger template
```
Comment (radtriste) [inline docs/PIPELINE.md:77]:
```suggestion

2) Create the trigger template
```

File: docs/PIPELINE.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+tkn el -n kogito-serverless-operator-pipeline ls
+```
+
+```sh
+kubectl get pods,svc -n kogito-serverless-operator-pipeline -l eventlistener=kogito-serverless-operator-webhook
+```
+
+8) Add a webhook in your github/gitlab repo with the url of the listener on openshift
+
+
```
Comment (radtriste) [inline docs/PIPELINE.md:113]:
```suggestion
```

**Review by radtriste — COMMENTED** [review]
Mostly formatting comments

File: ci/PIPELINE.md [hunk 15]
```diff
@@ -0,0 +1,117 @@
+## Install and configure a Tekton pipeline on Openshift
+
+1) Install Red Hat OpenShift Pipelines latest (1.9.2 at the time of this doc)  from the Operator Hub
+
+2) If you want interact via cli with the pipeline you can install locally Tekton cli 
+   to interact with the pipelines, you can download from the details of the operator installed (i.e. crc link)
```
[the remark this answers is not in the document]
Comment (spolti) [reply ci/PIPELINE.md]: can you please link it here?

File: ci/PIPELINE.md (hunk 9 above)
Comment (spolti) [reply ci/PIPELINE.md]: maybe add a print as example/guidance?

File: docs/PIPELINE.md [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+1) Create the task show_workspace_content
+
+```sh
+kubectl apply -f tekton/task/show_workspace_content.yaml
+```
+
+2) Add the task show-workspace in the pipeline after the fetch-repository or build-image
+
+### How to deploy again
```
Comment (spolti) [inline docs/PIPELINE.md:66]:
```suggestion
### How to redeploy
```

File: tekton/role/cluster_role.yaml [hunk 17]
```diff
@@ -0,0 +1,152 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: tekton-clustermanger-role
+rules:
+- apiGroups:
+    - rbac.authorization.k8s.io
+  resources:
```
[the remark this answers is not in the document]
Comment (desmax74) [reply tekton/role/cluster_role.yaml:8]: Yep, added to fix one error at time during the start of the operator :/

**Review by radtriste — APPROVED** [review]

File: tekton/role/cluster_role.yaml (hunk 17 above)
[the remark this answers is not in the document]
Comment (spolti) [reply tekton/role/cluster_role.yaml:8]: I think we should fine tune this, there is a lot of high privilege roles here.

[the remark this answers is not in the document]
Comment (spolti) [reply tekton/role/cluster_role.yaml:8]: example, create and delete nodes, service accounts, clusterroles, cluster role bindings...

**Pull request merged**